src/audio/google/google_rtc_audio_processing.c

```diff
@@ -60,53 +83,58 @@ struct google_rtc_audio_processing_comp_data {
 	int num_aec_reference_channels;
 	int num_capture_channels;
 	GoogleRtcAudioProcessingState *state;
-	int16_t *aec_reference_buffer;
-	int aec_reference_frame_index;
-	int16_t *raw_mic_buffer;
-	int raw_mic_buffer_frame_index;
-	int16_t *output_buffer;
-	int output_buffer_frame_index;
-	uint8_t *memory_buffer;
+	float *raw_mic_buffers[MIC_CHAN_MAX];
+	float *refout_buffers[REFOUT_CHAN];
+	int buffered_frames;
 	struct comp_data_blob_handler *tuning_handler;
 	bool reconfigure;
 	int aec_reference_source;
 	int raw_microphone_source;
+	struct comp_buffer *ref_comp_buffer;
+	int ref_frame_bytes;
+	int out_frame_bytes;
 };
 
-void *GoogleRtcMalloc(size_t size)
+#if CONFIG_GOOGLE_RTC_AUDIO_PROCESSING_MIC_BITS == 16
+typedef int16_t mic_sample_t;
+#define MIC_SCALE ((float)SHRT_MAX)
```

Comment:
SHRT_MAX  + 1

Reply:
Is that correct?  Doing it that way would cause the value 1.0 to produce a 0x8000 on conversion back to integer, which is SHRT_MIN and thus emits a -1.0 and not 1.0.  We'd need a saturation check in the path to avoid this.

The way the code here works is to represent [-1.0f:1.0f] as [-0x7fff:0x7fff], which is symmetric.

Or maybe this is a collision between assembly hacker brains and signal processor brains?  Is this just not the right convention per industry practice?

Reply:
When converting to int16, one should saturate:
```
[](float x) { return static_cast<int16_t>(std::clamp(x * (1 << 15), (float) std::numeric_limits<int16_t>::min(), (float) std::numeric_limits<int16_t>::max()));
```
This is in c++ but you get the idea.

Reply:
Lionel set me straight offline, fixed